Sources/GRPC/ClientConnectivityHandler.swift

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import NIOHTTP2
+
+internal class ClientConnectivityHandler: ChannelInboundHandler {
+  typealias InboundIn = HTTP2Frame
+
+  private var connectionManager: ConnectionManager
+  private let idleTimeout: TimeAmount
+
+  private var activeStreams = 0
+  private var scheduledIdle: Scheduled<Void>? = nil
+  private var state: State = .notReady
+
+  private enum State {
+    // We haven't marked the connection as "ready" yet.
+    case notReady
+
+    // The connection has been marked as "ready".
+    case ready
+
+    // We called `close` on the channel.
+    case closed
+  }
+
+  init(connectionManager: ConnectionManager, idleTimeout: TimeAmount = .minutes(5)) {
+    self.connectionManager = connectionManager
+    self.idleTimeout = idleTimeout
+  }
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
```

Comment:
This should probably also guard on whether you've called `close`.

Reply:
done